portal/src/util/passwordGenerator.ts

```diff
@@ -0,0 +1,182 @@
+    return generatePasswordWithSource(mathRandSource, policy);
+  }
+}
+
+export function generatePasswordWithSource(
+  source: RandSource,
+  policy: PasswordPolicyConfig
+): string | null {
+  const [password, _] = internalGeneratePasswordWithSource(source, policy);
+  return password;
+}
+
+export function internalGeneratePasswordWithSource(
+  source: RandSource,
+  policy: PasswordPolicyConfig
+): [string | null, number] {
+  for (let i = 0; i < MaxTrials; i++) {
+    const password = generatePasswordOnce(source, policy);
+    if (password !== null) {
+      return [password, i];
+    }
+  }
+
+  return [null, -1];
+}
+
+// eslint-disable-next-line complexity
+function generatePasswordOnce(
+  source: RandSource,
+  policy: PasswordPolicyConfig
+): string | null {
+  const minLength = determineMinLength(policy);
+  const charList = prepareCharList(policy);
+
+  const passwordArray: string[] = [];
+
+  // Add required characters.
+  if (policy.lowercase_required) {
+    passwordArray.push(pickRandChar(source, CharListLowercase));
+  }
+  if (policy.uppercase_required) {
+    passwordArray.push(pickRandChar(source, CharListUppercase));
+  }
+  if (
+    policy.alphabet_required &&
+    !policy.lowercase_required &&
+    !policy.uppercase_required
+  ) {
+    passwordArray.push(pickRandChar(source, CharListAlphabet));
+  }
+  if (policy.digit_required) {
+    passwordArray.push(pickRandChar(source, CharListDigit));
+  }
+  if (policy.symbol_required) {
+    passwordArray.push(pickRandChar(source, CharListSymbol));
+  }
+
+  // Fill the rest of the password with random characters.
+  for (let i = passwordArray.length; i < minLength; i++) {
+    passwordArray.push(pickRandChar(source, charList));
+  }
+
+  // Shuffle the password since we have required characers at the beginning.
+  for (let i = passwordArray.length - 1; i > 0; i--) {
+    const j = source.intN(i + 1);
+    [passwordArray[i], passwordArray[j]] = [passwordArray[j], passwordArray[i]];
+  }
+
+  const password = passwordArray.join("");
+  if (validatePassword(password, policy) === null) {
+    return password;
+  }
+
+  return null;
+}
+
+export function prepareCharList(passwordPolicy: PasswordPolicyConfig): string {
+  const set = new Set<string>();
+
+  if (passwordPolicy.alphabet_required) set.add(CharListAlphabet);
+  if (passwordPolicy.lowercase_required) set.add(CharListLowercase);
+  if (passwordPolicy.uppercase_required) set.add(CharListUppercase);
+  if (passwordPolicy.digit_required) set.add(CharListDigit);
+  if (passwordPolicy.symbol_required) set.add(CharListSymbol);
+
+  // Default to alphanumeric if no character set is required.
+  if (set.size === 0) {
+    set.add(CharListAlphabet);
+    set.add(CharListDigit);
+  }
+
+  // Remove overlapping character sets.
+  if (set.has(CharListAlphabet)) {
+    set.delete(CharListLowercase);
+    set.delete(CharListUppercase);
+  }
+
+  // Build the final character list.
+  let charList = "";
+  set.forEach((cs) => {
```

Comment:
This does not match the server implementation. The server implementation uses allCharacterSets list to ensure a predictable order.

Reply:
Rebased and updated in [c4eabcd](https://github.com/authgear/authgear-server/pull/4537/commits/c4eabcdcafbc2f50cf4410c3e900a62300bb66ff)